### PR TITLE
feat: add end2end tests for practice epic

### PR DESCRIPTION
**Pull request by SaFouNaldo**
- This PR includes updates to address dependencies issues, introduces an offline mode feature, adds tests, and creates the first end-to-end test for the app.

## Dependency Update:

- Updated the build.gradle.kts to resolve dependency issues preventing the usage of createAndroidComposeRule for UI tests. - The issue was related to protobuf-related dependencies.
## Offline Mode Feature:

- Implemented an offline mode where only the "Practice Epic" is accessible.
- All other user-related screens display a white screen prompting the user to log in to continue.
- The white offline screen utilizes common elements from the utils package for consistent UI design.
- Added unit tests for the white offline screen to verify the displayed elements and functionality.
## End-to-End Test:

- Added the first end-to-end test, which navigates from the app's start to the ASL Recognition screen to recognize the word "dog".
## Future Work:

- Extend the end-to-end test to input hand gestures (D-O-G) into the camera feed for fully simulating a lambda user's behavior.
## Challenges

- The task required extensive debugging, especially related to protobuf dependency issues, resulting in a total development time of 15+ hours.
- Creating the end-to-end test was particularly challenging due to user dependencies across different screens. These dependencies created conflicts during navigation. Implementing the offline mode simplified these dependencies significantly, resolving two related issues.
## Testing

- Verified the offline mode screens display the correct UI and functionality in various states (offline, navigating to login).
- Validated the end-to-end test flow from app start to the ASL Recognition feature.

This PR lays the foundation for further enhancements in offline functionality and comprehensive end-to-end testing.

### CODE REVIEW

File: app/src/main/java/com/github/se/signify/ui/screens/LoginScreen.kt
Comment (NBCCF025) [inline app/src/main/java/com/github/se/signify/ui/screens/LoginScreen.kt:1]: Is it necessary to have the following line "isOfflineState = false // Set offline state to false when signed in" both in line 72 and 136. Shouldn't it be set only when the authentication is complete (when the toast of a successful login is send) at line 72 ?

**Review by NBCCF025 — CHANGES_REQUESTED** [review]
#### Overview
This PR demonstrates significant progress by addressing dependency issues, introducing an offline mode, and creating the first end-to-end test for the app. The changes are well-documented and provide a solid foundation for future improvements.

---

#### **End To End Test**
- The first end to end test is well implemented and correctly test the practice epic, but we should remind us that we used the offline mode for this one and bypass the login logic.

#### **Offline Mode Feature**
- This feature is great and provide the following result (images of the screens):

login
ProfileChallenge


---

#### **Requested Changes**
- The `isOfflineState` shouldn't be put as false just by clicking on the button to connect, it should be assigned only after a successful login.
- Offline mode need some changes. I currently tested it with and whitout internet connectivity, and I have seen the Following:
	1) (with internet connectivity):
		- The feedback are accessible, can be filled and can be send to the firebase repository as it doesn't require a connected user.
		- The quiz screen is accessible but doesn't show anything except the title and the back button, but I don't know why as the quizzes doesn't require a connected user and I have internet connectivity in this case to access the firebase repository.
		- The quest screen is accessible but doesn't show anything except the title and the back button as there is no connected user to access the `initialQuestAccessDate`.
	2) (without internet connectivity):
		- The feedback are accessible, can be filled but can't be send to the firebase repository (The last one is normal as I have no internet connectivity).
		- The quiz screen is accessible but doesn't show anything except the title and the back button as there is no internet connectivity to access the quiz repository.
		- The quest screen is accessible but doesn't show anything except the title and the back button as there is no connected user and no internet connectivity to access the firebase repository.
	- I suggest that this 3 screens (feedbackScreen, QuizScreen and QuestScreen) take a similar behaviour as ProfileScreen and ChallengeScreen for offline mode, so it requires the user to be connected.

homeTop
feedback
firebaseFeedback
quiz
quest


---

### Final Comments
This PR successfully addresses its purpose, but I recommand some changes before merging it.
Great work overall!

**Comment by SaFouNaldo** [comment]
> #### Overview
> This PR demonstrates significant progress by addressing dependency issues, introducing an offline mode, and creating the first end-to-end test for the app. The changes are well-documented and provide a solid foundation for future improvements.
> 
> #### **End To End Test**
> * The first end to end test is well implemented and correctly test the practice epic, but we should remind us that we used the offline mode for this one and bypass the login logic.
> 
> #### **Offline Mode Feature**
> * This feature is great and provide the following result (images of the screens):
> 
> login ProfileChallenge
> 
> #### **Requested Changes**
> * The `isOfflineState` shouldn't be put as false just by clicking on the button to connect, it should be assigned only after a successful login.
> * Offline mode need some changes. I currently tested it with and whitout internet connectivity, and I have seen the Following:
>   
>   1. (with internet connectivity):
>      
>      * The feedback are accessible, can be filled and can be send to the firebase repository as it doesn't require a connected user.
>      * The quiz screen is accessible but doesn't show anything except the title and the back button, but I don't know why as the quizzes doesn't require a connected user and I have internet connectivity in this case to access the firebase repository.
>      * The quest screen is accessible but doesn't show anything except the title and the back button as there is no connected user to access the `initialQuestAccessDate`.
>   2. (without internet connectivity):
>      
>      * The feedback are accessible, can be filled but can't be send to the firebase repository (The last one is normal as I have no internet connectivity).
>      * The quiz screen is accessible but doesn't show anything except the title and the back button as there is no internet connectivity to access the quiz repository.
>      * The quest screen is accessible but doesn't show anything except the title and the back button as there is no connected user and no internet connectivity to access the firebase repository.
>   
>   
>   * I suggest that this 3 screens (feedbackScreen, QuizScreen and QuestScreen) take a similar behaviour as ProfileScreen and ChallengeScreen for offline mode, so it requires the user to be connected.
> 
> homeTop feedback firebaseFeedback quiz quest
> 
> ### Final Comments
> This PR successfully addresses its purpose, but I recommand some changes before merging it. Great work overall!

Hello, Thanks for your detailed review : 
- I absolutely agree with the fact to take the similiar behaviour as ProfileScreen in the screens like Quizz, feedback and Quests and I will updated it shortly.
- >The isOfflineState shouldn't be put as false just by clicking on the button to connect, it should be assigned only after a >successful login.
Well I think you missunderstood how the composable works : the isOfflineState is changing only when the Authetication is complete (onAuth = )

File: app/src/androidTest/java/com/github/se/signify/ui/screens/home/FeedbackScreenTest.kt
Comment (NBCCF025) [inline app/src/androidTest/java/com/github/se/signify/ui/screens/home/FeedbackScreenTest.kt:1]: We should have a `setUp()` function for the `setContent` on this file test, but as it is not related to the change done in this PR, we can let it like that.

**Review by NBCCF025 — APPROVED** [review]
It is all good now, this PR is approved.

But you said that I missunderstood the logic with the `isOfflineState`. However, you removed the duplicate from the GoogleSignInButton to only let the one in the onAuthComplete. And this was what I meant by the change to do.

**Pull request merged**